tests/rptest/tests/full_disk_test.py

```diff
@@ -0,0 +1,118 @@
+        _get = lambda k: self.admin.get_cluster_config(node=node)[k]
+
+        def match(key: str, val: int) -> bool:
+            v = _get(key)
+            self.logger.info(
+                f"config wait: want {type(value)} {value}, got {v}" +
+                f" {type(v)}")
+            return v == value
+
+        wait_until(lambda: match(key, value), timeout_sec=15)
+        self.logger.debug(f"Verified val {value} for {key}.")
+
+    def _trigger_low_space(self):
+        victim = random.choice(self.redpanda.nodes)
+        vfree = self.node_metrics.disk_free_bytes_node(victim)
+        new_threshold = int(max(vfree * 2, 1024))
+        updates = {self.CONF_MIN_FREE_BYTES: new_threshold}
+        self.logger.debug(f" victim #{victim}, vfree {vfree} " +
+                          f"new_thresh {new_threshold}")
+        self.admin.patch_cluster_config(upsert=updates, remove=[])
+
+        self.logger.info(f"Confirming new config values..")
+        self._wait_for_node_config_value(victim, self.CONF_MIN_FREE_BYTES,
+                                         new_threshold)
+
+    def _setup(self) -> None:
+        self.logger.debug(f"brokers: {self.redpanda.brokers_list()}")
+        self.producer = KafkaProducer(
+            acks="all",
+            bootstrap_servers=self.redpanda.brokers_list(),
+            value_serializer=str.encode)
+        self.node_metrics = NodeMetrics(self.redpanda)
+        self.admin = Admin(self.redpanda)
+
+        # XXX TODO wait_for() something instead?
+        # To avoid "NotLeaderForPartitionError"
+        sleep(5)
+
+    @cluster(num_nodes=N, log_allow_list=FDT_LOG_ALLOW_LIST)
+    def test_write_rejection(self):
+        """ Verify that we block external writers when free disk space on any
+        node is below the threshold. """
+
+        t0 = time()
+        self._setup()
+        t1 = time()
+
+        # 1. Confirm we can produce successfully.
+        future = self.producer.send(self.TOPIC, "Q: What do redpandas eat?")
+        # TODO async version of wait_until() would be sweet.
+        self.node_metrics.wait_until_ready()
+        r = future.get(timeout=15)
+        self.logger.debug("Produce result RecordMetadata {r}")
+
+        # 2. Set threshold below current free space, assert we get rejected.
+        t2 = time()
+        self._trigger_low_space()
+        t3 = time()
+        was_blocked = False
+        pause = 1.0 / MAX_MSG_PER_SEC
+        for i in range(MAX_MSG):
+            try:
+                future = self.producer.send(
+                    self.TOPIC,
+                    "A: Don't know, but they can do a lot of bites per second."
+                )
+                sleep(pause)
+                future.get(timeout=15)
+                self.logger.debug(f"sent msg {i}/{MAX_MSG}, t={time()-t3:.2f}")
+            except BaseException as e:
+                self.logger.info(f"Write rejected as expected: {e}")
+                was_blocked = True
```

Comment:
how can we test for the specific out-of-space scenario as the cause for the produce failing, and exclude (as much as possible) all the other reasons a produce might fail?

Reply:
See updated patch (just force pushed).